service.go

```diff
@@ -55,132 +53,120 @@ func (s *Service) tickerService() {
-	dev, err := s.getDeviceByID(id)
+func (s *Service) UpdateDeviceInterval(id int, interval int) error {
+	dev, err := s.GetDeviceByID(id)
 	if err != nil {
 		return err
 	}
 	dev.Interval = interval
 	return nil
 }
 
-func (s *Service) updateDeviceValue(d *Device, value string) {
+func (s *Service) updateDeviceValue(d *Device, value float64) {
 	d.Value = value
 }
 
-func (s *Service) startDevice(d *Device, getMeasurement measurement) error {
-	err := s.addDevice(d)
-	if err != nil {
-		return err
-	}
-	go d.deviceTicker(s, getMeasurement)
-	return nil
-}
-
-func (s *Service) addDevice(d *Device) error {
-	if s.deviceAlreadyExists(d.Id) {
-		err := errors.New("The device with given ID already exists!")
-		return err
-	}
-	s.devices[d.Id] = *d
-	return nil
-}
-
 func (s *Service) stopDevice(d *Device) {
 	fmt.Printf("Stopping %s ID:%d...\n", d.Name, d.Id)
 	d.stopChan <- true
-	s.removeDevice(d)
 }
 
-func (s *Service) removeDevice(d *Device) {
-	delete(s.devices, d.Id)
+func (s *Service) RemoveDevice(d *Device) {
+	s.stopDevice(d)
+	delete(s.dao.Devices, d.Id)
 	time.Sleep(50 * time.Millisecond)
 	fmt.Printf("%s ID:%d removed.\n", d.Name, d.Id)
 }
 
-func (s *Service) getDevicesList() {
-	for _, dev := range s.devices {
+func (s *Service) GetDevicesList() {
+	for _, dev := range s.dao.Devices {
 		fmt.Printf("%s -- ID:%d and interval=%d mls\n",
 			dev.Name, dev.Id, dev.Interval)
 	}
 }
 
-func (s *Service) getDeviceByID(id int) (*Device, error) {
+func (s *Service) GetDeviceByID(id int) (*Device, error) {
 	var err error
 	if s.deviceAlreadyExists(id) {
-		dev := s.devices[id]
+		dev := s.dao.Devices[id]
 		return &dev, nil
 	}
 	err = errors.New("The device with given ID doesn't exist!")
 	return nil, err
 }
 
 func (s *Service) deviceAlreadyExists(id int) bool {
-	for _, dev := range s.devices {
+	for _, dev := range s.dao.Devices {
 		if dev.Id == id {
 			return true
 		}
 	}
 	return false
 }
 
-func (s *Service) getReadings() []string {
+func (s *Service) GetReadings() []string {
 	var fwdReadings []string
-	for device, readings := range s.DevicesReadings.data {
+	for device, readings := range s.dao.Readings {
 		fwdReadings = append(fwdReadings, fmt.Sprintf("Device ID:%d\n", device))
 		for _, r := range readings {
 			fwdReadings = append(fwdReadings,
-				fmt.Sprintf("Nanoseconds: %d -- with value %s\n", r.When.Nanosecond(), r.Value))
+				fmt.Sprintf("Nanoseconds: %d -- with value %f\n", r.When.Nanosecond(), r.Value))
 		}
 	}
 	return fwdReadings
 }
+
+func (s *Service) validate(payload *DevicePayload) error {
+	validate := validator.New()
+	validationErrors := validate.Struct(payload)
+	if validationErrors != nil {
+		for _, err := range validationErrors.(validator.ValidationErrors) {
+			fmt.Println(err)
+		}
+		errForwarded := errors.New("input validation failed, device not created")
+		return errForwarded
+	}
+	return nil
+}
+
+func (s *Service) AddDevice(payload *DevicePayload) (int, error) {
```

Comment:
Maybe we have not understood each other. My idea was to return `(int, error)` from DAO but the service should return `(*Device, error)`.